src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -0,0 +1,581 @@
+            if (_size < 0)
+            {
+                throw new FormatException(string.Format(SR.TarSizeFieldNegative, _name));
+            }
+
+            // Continue with the rest of the fields that require no special checks
+
+            _name = TarHelpers.GetTrimmedUtf8String(buffer.Slice(FieldLocations.Name, FieldLengths.Name));
+            _mode = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(buffer.Slice(FieldLocations.Mode, FieldLengths.Mode));
+            _uid = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(buffer.Slice(FieldLocations.Uid, FieldLengths.Uid));
+            _gid = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(buffer.Slice(FieldLocations.Gid, FieldLengths.Gid));
+            int mTime = TarHelpers.GetTenBaseNumberFromOctalAsciiChars(buffer.Slice(FieldLocations.MTime, FieldLengths.MTime));
+            _mTime = TarHelpers.GetDateTimeFromSecondsSinceEpoch(mTime);
+            _typeFlag = (TarEntryType)buffer[FieldLocations.TypeFlag];
+            _linkName = TarHelpers.GetTrimmedUtf8String(buffer.Slice(FieldLocations.LinkName, FieldLengths.LinkName));
+
+            if (_format == TarFormat.Unknown)
+            {
+                _format = _typeFlag switch
+                {
+                    TarEntryType.ExtendedAttributes or
+                    TarEntryType.GlobalExtendedAttributes => TarFormat.Pax,
+
+                    TarEntryType.DirectoryList or
+                    TarEntryType.LongLink or
+                    TarEntryType.LongPath or
+                    TarEntryType.MultiVolume or
+                    TarEntryType.RenamedOrSymlinked or
+                    TarEntryType.SparseFile or
+                    TarEntryType.TapeVolume => TarFormat.Gnu,
+
+                    // V7 is the only one that uses 'V7RegularFile'.
+                    TarEntryType.V7RegularFile => TarFormat.V7,
+
+                    // We can quickly determine the *minimum* possible format if the entry type
+                    // is the POSIX 'RegularFile', although later we could upgrade it to PAX or GNU
+                    _ => (_typeFlag == TarEntryType.RegularFile) ? TarFormat.Ustar : TarFormat.V7
+                };
+            }
+
+            return true;
+        }
+
+        // Reads fields only found in ustar format or above and converts them to their expected data type.
+        // Throws if any conversion fails.
+        private void ReadMagicAttribute(Span<byte> buffer)
+        {
+            Span<byte> magic = buffer.Slice(FieldLocations.Magic, FieldLengths.Magic);
+
+            // If at this point the magic value is all nulls, we definitely have a V7
+            if (TarHelpers.IsAllNullBytes(magic))
+            {
+                _format = TarFormat.V7;
+                return;
+            }
+
+            // When the magic field is set, the archive is newer than v7.
+            _magic = Encoding.ASCII.GetString(magic);
```

Comment:
I assume this "magic" is typically only one of a few known values?  Can we special-case those values to use a string literal rather than allocating a new one each time?

Reply:
Sure no problem.